dbms/src/Common/ConcurrentIOQueue.h

```diff
@@ -26,132 +26,127 @@ namespace DB
   * while (queue.isFull()) {}
   * queue.nonBlockingPush(std::move(obj));
   * ```
-  * There are additional overheads compared to using mpmcqueue directly
-  * - two locks
-  * - condition double check
-  * Since io queues are not performance sensitive, they are acceptable.
   */
 template <typename T>
 class ConcurrentIOQueue
 {
 public:
-    explicit ConcurrentIOQueue(size_t capacity)
-        : mpmc_queue(std::max(1, capacity))
-    {}
-
-    ConcurrentIOQueue(
-        size_t capacity,
-        Int64 max_auxiliary_memory_usage_,
-        typename MPMCQueue<T>::ElementAuxiliaryMemoryUsageFunc && get_auxiliary_memory_usage_)
-        : mpmc_queue(std::max(1, capacity), max_auxiliary_memory_usage_, std::move(get_auxiliary_memory_usage_))
+    explicit ConcurrentIOQueue(size_t capacity_)
+        : capacity(std::max(1, capacity_))
     {}
 
     /// blocking function.
     /// Just like MPMCQueue::push.
     MPMCQueueResult push(T && data)
     {
-        return mpmc_queue.push(std::move(data));
+        std::unique_lock lock(mu);
+        writer_head.wait(lock, [&] { return queue.size() < capacity || (unlikely(status != MPMCQueueStatus::NORMAL)); });
+
+        if ((likely(status == MPMCQueueStatus::NORMAL)) && queue.size() < capacity)
+        {
+            pushFront(std::move(data));
+            return MPMCQueueResult::OK;
+        }
+
+        switch (status)
+        {
+        case MPMCQueueStatus::NORMAL:
+            return MPMCQueueResult::FULL;
+        case MPMCQueueStatus::CANCELLED:
+            return MPMCQueueResult::CANCELLED;
+        case MPMCQueueStatus::FINISHED:
+            return MPMCQueueResult::FINISHED;
+        }
     }
 
     /// Just like MPMCQueue::tryPush.
     template <typename U>
     MPMCQueueResult tryPush(U && data)
     {
-        return mpmc_queue.tryPush(std::forward<U>(data));
+        std::lock_guard lock(mu);
+
+        if unlikely (status == MPMCQueueStatus::CANCELLED)
+            return MPMCQueueResult::CANCELLED;
+        if unlikely (status == MPMCQueueStatus::FINISHED)
+            return MPMCQueueResult::FINISHED;
+
+        if (queue.size() >= capacity)
+            return MPMCQueueResult::FULL;
+
+        pushFront(std::forward<U>(data));
+        return MPMCQueueResult::OK;
     }
 
     /// Non-blocking function.
     /// Besides all conditions mentioned at `push`, `nonBlockingPush` will still return OK if queue is `NORMAL` and full.
-    /// The obj that exceeds its capacity will be stored in remaining_objs and wait for the next pop/tryPop to trigger pushRemains.
+    /// The obj that exceeds its capacity will still be stored in queue.
     MPMCQueueResult nonBlockingPush(T && data)
```

Comment:
Maybe the name `forcePush` is more proper.

Reply:
good idea!
I renamed all `nonBlockingPush` and `nonBlockingWrite` to `forcePush` and `forceWrite`